include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -181,18 +174,56 @@ class pool_memory_resource final
    */
   block_type expand_pool(size_t size, free_list& blocks, cudaStream_t stream)
   {
-    auto grow_size = size_to_grow(size);
-    if (grow_size == 0) {
-      RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
-                    reinterpret_cast<void*>(stream),
-                    size);
-      RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
+    // Strategy. If maximum_pool_size_ is set, then grow geometrically, e.g. by halfway to the limit
+    // each time. If it is not set, grow exponentially, e.g. by doubling the pool size each time.
+    // Upon failure, attempt to back off exponentially, e.g. by half the attempted size, until
+    // either success or the attempt is less than the requested size.
+    auto try_size = size_to_grow(size);
+    while (try_size >= size) {
+      try {
+        auto b = block_from_upstream(try_size, stream);
```

Comment:
Might add a comment like this to make it clear why there is a `+= try_size` happening after `b = ` and not before.
```suggestion
        auto b = block_from_upstream(try_size, stream);  // has side-effect: can throw
```

Reply:
Since we're already using `thrust::optional`, maybe `block_from_upstream` can be made to return a `thrust::optional<block>`?

Reply:
`thrust::optional` is better than exception throwing IMO, but I would prefer the `expected`/`Result` type: https://github.com/rapidsai/rmm/pull/601#discussion_r504104258

Reply:
No matter what, the upstream memory resource can throw an exception.

Reply:
> No matter what, the upstream memory resource can throw an exception.

Yeah, the idea would be to catch the thrown exception inside `block_from_upstream` and return a `optional<block>` whose contents depend on the success of the allocation from the Upstream.